### PR TITLE
Rust: Include patterns as data flow nodes

### PR DESCRIPTION
**Pull request by paldepind**
Adds patterns as data flow nodes.

Also adds flow from the rhs. of a `let` statement to the lhs. which depends on patterns being in the data flow nodes. That will need some tweaking once we have the CFG layer over the AST layer.

Adds an abstract class for data flow node that corresponds to a CFG nodes for a AST nodes to factor out some common code.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -123,6 +123,19 @@
     override Location getLocation() { none() }
   }
 
+  /** A data flow node that corresponds to a CFG node for an AST node. */
+  abstract private class AstCfgFlowNode extends Node {
```
[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:127]: You may be able to suppress it with something like
```ql
    bindingset[this, n]
    AstCfgFlowNode() { exists(this) and exists(n) }
```

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -281,6 +289,11 @@ module LocalFlow {
   pragma[nomagic]
   predicate localFlowStepCommon(Node nodeFrom, Node nodeTo) {
     nodeFrom.getCfgNode() = getALastEvalNode(nodeTo.getCfgNode())
+    or
+    exists(LetStmt s |
+      nodeFrom.getCfgNode().getAstNode() = s.getInitializer() and
+      nodeTo.getCfgNode().getAstNode() = s.getPat()
+    )
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:296]: This does not take CFG splitting into account, but that changes once https://github.com/github/codeql/pull/17918 lands.

**Review by hvitved — APPROVED** [review]
LGTM.

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
Comment (paldepind) [reply rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:296]: Yes. I can also adapt this after #17918 (if you don't want to do it all yourself).

**Pull request merged**